### PR TITLE
docs: Add 004 solution architecture ADR

### LINKED ISSUE #82: Document solution architecture ADR
Draw a high level diagram of all components involved in 10101.

coordinator, mobile app, order book, etc.

### PR DESCRIPTION
**Pull request by holzeis**
resolves #82

### CODE REVIEW

**Comment by holzeis** [comment]
I must say I am not a big fan of the github mermaid integration. The diagram looks completely different than how it looked in my editor. 🙁

File: docs/004-solution-architecture.adoc [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+The mobile app uses the order book to find trades and the coordinator to establish a "transient" dlc-channel with the maker respectively taker.
+
+=== Order book
+
+The order book is responsible for publishing orders and matching them accordingly. Once a match is found, the order book instructs the coordinator to establish the technical foundations to execute the trade between the taker and the maker.
+
+
+=== Coordinator
+
+The coordinator coordiantes the creation of the technical dlc channel and exposes an always one endpoint for the user to interact with her trade / dlc.
```
Comment (bonomat) [inline docs/004-solution-architecture.adoc:67]:
```suggestion
The coordinator coordiantes the creation of the technical dlc channel and exposes an always on endpoint for the user to interact with her trade / dlc.
```

File: docs/004-solution-architecture.adoc [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        ContainerDb_Ext(mobile_db, "Database", "SqLite", "Manages the domain, reference, technical and sensible data for the phone")
+    }
+
+    Boundary(b2, "Coordinator", "boundary") {
+        Container(coordinator, "Coordinator", "Rust, Lightning Node", "Coordinates the dlc-channel management")
+        ContainerDb(coordinator_db, "Database", "PostgreSQL", "Manages the technical and sensible data for the coordiantor")
+    }
+
+    Boundary(b3, "Orderbook", "boundary") {
+        Container(orderbook, "Order book", "Rust", "Matches maker and taker orders")
```
Comment (bonomat) [inline docs/004-solution-architecture.adoc:33]: 🙃
```suggestion
        Container(orderbook, "Orderbook", "Rust", "Matches maker and taker orders")
```

File: docs/004-solution-architecture.adoc [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+Rel(coordinator, coordinator_db, "Uses")
+Rel(orderbook, orderbook_db, "Uses")
+```
+
+=== Mobile App
+
+The mobile app exposes the lightning and on-chain wallet combined with a non-custodial trading functionality to the user. Using the mobile app the user runs a full lightning node on her phone.
+
+The mobile app uses the order book to find trades and the coordinator to establish a "transient" dlc-channel with the maker respectively taker.
```
Comment (bonomat) [inline docs/004-solution-architecture.adoc:58]:
```suggestion
The mobile app uses the orderbook to find trades and the coordinator to establish a "transient" dlc-channel with the maker respectively taker.
```

File: docs/004-solution-architecture.adoc [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+=== Order book
```
Comment (bonomat) [inline docs/004-solution-architecture.adoc:60]: 🙃
```suggestion
=== Orderbook
```

File: docs/004-solution-architecture.adoc [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+The order book is responsible for publishing orders and matching them accordingly. Once a match is found, the order book instructs the coordinator to establish the technical foundations to execute the trade between the taker and the maker.
```
Comment (bonomat) [inline docs/004-solution-architecture.adoc:62]: 🙃
```suggestion
The orderbook is responsible for publishing orders and matching them accordingly. Once a match is found, the orderbook instructs the coordinator to establish the technical foundations to execute the trade between the taker and the maker.
```

**Review by bonomat — APPROVED** [review]
Looks good. Thanks for writing this up.

> I must say I am not a big fan of the github mermaid integration. The diagram looks completely different than how it looked in my editor. 🙁

The code also looks kinda ugly 🤢🙈 Thanks for giving this a try.

**Comment by da-kami** [comment]
> I must say I am not a big fan of the github mermaid integration. The diagram looks completely different than how it looked in my editor. 🙁

It does indeed look ugly 😔

We can opt for `plantuml` after all. 

Rendering the image (through the plantuml server) is a bit more cumbersome then. We have to use the trick of adding the plantuml file first and then configure the plantuml server URL in the doc so we get the rendered image back for display.

This can also be seen as advantage to some extent -  the diagram will live in a separate file and we can reference it from different files in the docs.

File: docs/004-solution-architecture.adoc [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+Rel(mobile_app, coordinator, "Establishes dlc channels with", "TCP, P2P Lightning")
+UpdateRelStyle(mobile_app, coordinator, $offsetY="0", $offsetX="-40")
+
+Rel(coordinator, coordinator_db, "Uses")
+Rel(orderbook, orderbook_db, "Uses")
+```
+
+=== Mobile App
+
+The mobile app exposes the lightning and on-chain wallet combined with a non-custodial trading functionality to the user. Using the mobile app the user runs a full lightning node on her phone.
```
Comment (da-kami) [inline docs/004-solution-architecture.adoc:56]:
```suggestion
The mobile app exposes the lightning and on-chain wallet combined with a self-custodial trading functionality to the user. Using the mobile app the user runs a full lightning node on her phone.
```

File: docs/004-solution-architecture.adoc (hunk 5 above)
Comment (da-kami) [inline docs/004-solution-architecture.adoc:62]:
```suggestion
The order book is responsible for publishing orders and matching them accordingly. Once a match is found, the order book instructs the coordinator to establish the technical foundations to execute the trade between the matched trading parties.
```

Comment (da-kami) [reply docs/004-solution-architecture.adoc:62]: I would try to move away from the "maker and taker" terminology.

File: docs/004-solution-architecture.adoc (hunk 1 above)
Comment (da-kami) [reply docs/004-solution-architecture.adoc:67]: 🙃 I would add a dash here to make it even more explicit for the reading flow:`always-on`

**Review by da-kami — APPROVED** [review]
Sorry for the Mermaid rabbi-hole - given that it is super ugly I would consider going back to using plantuml for the diagram 😬

File: docs/004-solution-architecture.adoc (hunk 1 above)
Comment (luckysori) [reply docs/004-solution-architecture.adoc:67]: I think it's not just the reading flow, it's grammatically incorrect: https://www.grammarly.com/blog/hyphen/.

> Generally, you need the hyphen only if the two or more words are functioning together as an adjective _before_ the noun they’re describing. If the noun comes first, leave the hyphen out.

**Comment by holzeis** [comment]
> Sorry for the Mermaid rabbi-hole - given that it is super ugly I would consider going back to using plantuml for the diagram 😬

We probably going to update that diagram anyways at some point. I suggest we merge this and next time we need to touch that ADR, we can switch to another tooling.

**Pull request merged**